### PR TITLE
use Urls instead of strings

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[Safely join url paths](https://trello.com/c/rh78knoL/255-safely-join-url-paths)

File: services/src/util/parsing.rs [hunk 1]
```diff
@@ -53,3 +54,112 @@ where
 
     deserializer.deserialize_any(StringOrVec(PhantomData))
 }
+
+/// Deserialize a base URL by enforcing a trailing slash and then
```
Comment (michaelmattig) [inline services/src/util/parsing.rs:58]: incomplete comment?

File: services/src/util/parsing.rs [hunk 2]
```diff
@@ -53,3 +54,112 @@ where
...
+    let mut url_string = String::deserialize(deserializer)?;
+
+    if !url_string.ends_with('/') {
+        url_string.push('/');
+    }
+
+    Url::parse(&url_string).map_err(D::Error::custom)
+}
+
+/// Deserialize a base URL by enforcing a trailing slash and then
```
Comment (michaelmattig) [inline services/src/util/parsing.rs:72]: incomplete comment?

File: services/src/util/config.rs [hunk 3]
```diff
@@ -101,7 +103,8 @@ pub trait ConfigElement {
 #[derive(Debug, Deserialize)]
 pub struct Web {
     pub bind_address: String,
```
Comment (michaelmattig) [inline services/src/util/config.rs:105]: why is this not a url?

File: services/src/server.rs [hunk 4]
```diff
@@ -204,19 +202,19 @@ mod tests {
 
     async fn queries(shutdown_tx: Sender<()>) {
         let web_config: config::Web = get_config_element().unwrap();
-        let base_url = format!("http://{}", web_config.bind_address);
+        let base_url = Url::parse(&format!("http://{}", web_config.bind_address)).unwrap();
```
Comment (michaelmattig) [inline services/src/server.rs:205]: we need to enfore the trailing / here as well or not?

File: services/src/pro/server.rs [hunk 5]
```diff
@@ -114,12 +115,9 @@ pub async fn start_pro_server(
...
         "Starting server… {}",
-        format!(
-            "http://{}/",
-            web_config
-                .external_address
-                .unwrap_or(web_config.bind_address)
-        )
+        web_config
+            .external_address
+            .unwrap_or(Url::parse(&format!("http://{}/", web_config.bind_address))?)
```
Comment (michaelmattig) [inline services/src/pro/server.rs:120]: we need to enforce the trailing / here as well or not?

**Review by michaelmattig — CHANGES_REQUESTED** [review]

File: services/src/pro/server.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/server.rs:120]: it is checked that it is a valid socket address, so we can safely add the trailing slash

File: services/src/util/config.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply services/src/util/config.rs:105]: bind address is a socket address as discussed

File: services/src/util/parsing.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply services/src/util/parsing.rs]: 215c6e0

File: services/src/util/parsing.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply services/src/util/parsing.rs]: 215c6e0

File: services/src/server.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply services/src/server.rs:201]: c5b312d

File: services/src/pro/server.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/server.rs:116]: c5b312d

**Review by michaelmattig — APPROVED** [review]

**Pull request merged**